File: packages/shared/src/schemas/stock.ts
// The stock recount: what one run found and what the last one left behind.
// Quantities are thousandths of a unit, so every figure here is an integer
// the same way a centime figure is: a fractional one means the answer came
// from something other than the server this client was generated against.

import { z } from "zod";

import type { LastStockRecountDto } from "../generated/LastStockRecountDto";
import type { StockDriftDto } from "../generated/StockDriftDto";
import type { StockRecountDto } from "../generated/StockRecountDto";
import { day, exactInteger } from "./common";
import type { Assert, Matches } from "./drift";

export const stockDriftSchema = z.object({
  product_id: exactInteger,
  name: z.string(),
  cached_milli: exactInteger,
  ledger_milli: exactInteger,
  difference_milli: exactInteger,
}) satisfies z.ZodType<StockDriftDto>;
type _StockDrift = Assert<Matches<StockDriftDto, typeof stockDriftSchema>>;

export const stockRecountSchema = z.object({
  day,
  products_checked: exactInteger,
  drifts: z.array(stockDriftSchema),
}) satisfies z.ZodType<StockRecountDto>;
type _StockRecount = Assert<Matches<StockRecountDto, typeof stockRecountSchema>>;

/** `last_run_day` is null on a shop that has never recounted, which is what a
 *  file says before the daily loop has woken once. */
export const lastStockRecountSchema = z.object({
  last_run_day: day.nullable(),
  drifts: z.array(stockDriftSchema),
}) satisfies z.ZodType<LastStockRecountDto>;
type _LastStockRecount = Assert<Matches<LastStockRecountDto, typeof lastStockRecountSchema>>;

File: crates/api/src/dto/stock.rs
//! Stock recounts and the drift a recount found.
//!
//! Re-exported by `super`, so every path outside this folder is unchanged.

use super::*;

/// One product the recount put right: what the column said it had, what its
/// movements add up to, and the difference between them. The name travels
/// with the id because the panel is read by a person and it is what the log
/// stored, so a past run reads the same after a rename.
#[derive(Debug, Clone, Serialize, TS)]
#[ts(export_to = "StockDriftDto.ts")]
pub struct StockDriftDto {
    pub product_id: i32,
    pub name: String,
    pub cached_milli: i64,
    pub ledger_milli: i64,
    pub difference_milli: i64,
}

impl From<Drift> for StockDriftDto {
    fn from(d: Drift) -> Self {
        StockDriftDto {
            product_id: d.product_id,
            difference_milli: d.difference_milli(),
            name: d.name,
            cached_milli: d.cached_milli,
            ledger_milli: d.ledger_milli,
        }
    }
}

/// What one run of the recount found and did. `products_checked` is there so
/// an empty drift list reads as "nothing is wrong" rather than as "nothing
/// was looked at".
#[derive(Debug, Clone, Serialize, TS)]
#[ts(export_to = "StockRecountDto.ts")]
pub struct StockRecountDto {
    /// The day on the shop's calendar the run was marked under.
    pub day: String,
    pub products_checked: i64,
    pub drifts: Vec<StockDriftDto>,
}

impl From<Report> for StockRecountDto {
    fn from(r: Report) -> Self {
        StockRecountDto {
            day: r.day,
            // A count of this shop's products. The bound is unreachable on
            // any file a shop could have, and a number that is merely
            // bounded beats a refusal on a screen that is only reporting.
            products_checked: i64::try_from(r.checked).unwrap_or(i64::MAX),
            drifts: r.drifts.into_iter().map(StockDriftDto::from).collect(),
        }
    }
}

/// The last run as the file remembers it. `last_run_day` is null when the
/// shop has never recounted, which is what a file opened for the first time
/// says before the daily loop has woken once.
#[derive(Debug, Clone, Serialize, TS)]
#[ts(export_to = "LastStockRecountDto.ts")]
pub struct LastStockRecountDto {
    pub last_run_day: Option<String>,
    pub drifts: Vec<StockDriftDto>,
}

impl From<LastRecount> for LastStockRecountDto {
    fn from(l: LastRecount) -> Self {
        LastStockRecountDto {
            last_run_day: l.last_run_day,
            drifts: l.drifts.into_iter().map(StockDriftDto::from).collect(),
        }
    }
}
